File: Win8.xaml/Plugins/SystemTransportControlsPlugin.cs
using System;
using System.Linq;
using Windows.Media;
using Windows.UI.Xaml;

namespace Microsoft.PlayerFramework
{
    /// <summary>
    /// A plugin used to connect the Windows 8.1 system media transport controls with the current media.
    /// </summary>
    public class SystemTransportControlsPlugin : PluginBase
    {
        /// <summary>
        /// Gets the instance of the SystemMediaTransportControls being used.
        /// </summary>
        public SystemMediaTransportControls SystemControls { get; private set; }

        /// <inheritdoc /> 
        protected override bool OnActivate()
        {
            if (!Windows.ApplicationModel.DesignMode.DesignModeEnabled)
            {
                SystemControls = SystemMediaTransportControls.GetForCurrentView();
                SystemControls.ButtonPressed += SystemControls_ButtonPressed;
                RefreshFastForwardState();
                RefreshRewindState();
                RefreshStopState();
                RefreshPlayState();
                RefreshPauseState();
                RefreshNextState();
                RefreshPreviousState();
                RefreshPlaybackStatus();

                if (PlaylistPlugin != null)
                {
                    PlaylistPlugin.Playlist.CollectionChanged += Playlist_CollectionChanged;
                    PlaylistPlugin.CurrentPlaylistItemChanged += PlaylistPlugin_CurrentPlaylistItemChanged;
                }
                WireEvents(MediaPlayer.InteractiveViewModel);
                MediaPlayer.InteractiveViewModelChanged += MediaPlayer_InteractiveViewModelChanged;
                return true;
            }
            else return false;
        }

        /// <inheritdoc /> 
        protected override void OnDeactivate()
        {
            if (PlaylistPlugin != null)
            {
                PlaylistPlugin.Playlist.CollectionChanged -= Playlist_CollectionChanged;
                PlaylistPlugin.CurrentPlaylistItemChanged -= PlaylistPlugin_CurrentPlaylistItemChanged;
            }

            IsNextTrackEnabled = false;
            IsPreviousTrackEnabled = false;

            MediaPlayer.InteractiveViewModelChanged -= MediaPlayer_InteractiveViewModelChanged;
            UnwireEvents(MediaPlayer.InteractiveViewModel);

            SystemControls.ButtonPressed -= SystemControls_ButtonPressed;
            SystemControls = null;
        }

        void WireEvents(IInteractiveViewModel viewModel)
        {
            if (viewModel != null)
            {
                viewModel.CurrentStateChanged += Player_CurrentStateChanged;
                viewModel.IsStopEnabledChanged += Player_IsStopEnabledChanged;
                viewModel.IsRewindEnabledChanged += Player_IsRewindEnabledChanged;
                viewModel.IsFastForwardEnabledChanged += Player_IsFastForwardEnabledChanged;
                viewModel.IsPlayResumeEnabledChanged += Player_IsPlayResumeEnabledChanged;
                viewModel.IsPauseEnabledChanged += Player_IsPauseEnabledChanged;
                viewModel.IsSkipNextEnabledChanged += Player_IsSkipNextEnabledChanged;
                viewModel.IsSkipPreviousEnabledChanged += Player_IsSkipPreviousEnabledChanged;
            }
        }

        void UnwireEvents(IInteractiveViewModel viewModel)
        {
            if (viewModel != null)
            {
                viewModel.CurrentStateChanged -= Player_CurrentStateChanged;
                viewModel.IsStopEnabledChanged -= Player_IsStopEnabledChanged;
                viewModel.IsRewindEnabledChanged -= Player_IsRewindEnabledChanged;
                viewModel.IsFastForwardEnabledChanged -= Player_IsFastForwardEnabledChanged;
                viewModel.IsPlayResumeEnabledChanged -= Player_IsPlayResumeEnabledChanged;
                viewModel.IsPauseEnabledChanged -= Player_IsPauseEnabledChanged;
                viewModel.IsSkipNextEnabledChanged -= Player_IsSkipNextEnabledChanged;
                viewModel.IsSkipPreviousEnabledChanged -= Player_IsSkipPreviousEnabledChanged;
            }
        }

        bool isPreviousTrackEnabled;
        bool IsPreviousTrackEnabled
        {
            get { return isPreviousTrackEnabled; }
            set
            {
                if (isPreviousTrackEnabled != value)
                {
                    isPreviousTrackEnabled = value;
                    SystemControls.IsPreviousEnabled = isPreviousTrackEnabled;
                }
            }
        }

        bool isNextTrackEnabled;
        bool IsNextTrackEnabled
        {
            get { return isNextTrackEnabled; }
            set
            {
                if (isNextTrackEnabled != value)
                {
                    isNextTrackEnabled = value;
                    SystemControls.IsNextEnabled = isNextTrackEnabled;
                }
            }
        }

        /// <summary>
        /// Gets if a next track currently exists in the playlist.
        /// </summary>
        protected virtual bool NextTrackExists
        {
            get
            {
                return PlaylistPlugin != null && PlaylistPlugin.NextPlaylistItem != null;
            }
        }

        /// <summary>
        /// Gets if a previous track currently exists in the playlist.
        /// </summary>
        protected virtual bool PreviousTrackExists
        {
            get
            {
                return PlaylistPlugin != null && PlaylistPlugin.PreviousPlaylistItem != null;
            }
        }

        void PlaylistPlugin_CurrentPlaylistItemChanged(object sender, EventArgs e)
        {
            RefreshNextState();
            RefreshPreviousState();
        }

        void Playlist_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e)
        {
            RefreshNextState();
            RefreshPreviousState();
        }

        PlaylistPlugin PlaylistPlugin
        {
            get
            {
                return MediaPlayer.Plugins.OfType<PlaylistPlugin>().FirstOrDefault();
            }
        }

        void Player_CurrentStateChanged(object sender, RoutedEventArgs e)
        {
            RefreshPlaybackStatus();
        }

        private void RefreshPlaybackStatus()
        {
            switch (MediaPlayer.InteractiveViewModel.CurrentState)
            {
                case Windows.UI.Xaml.Media.MediaElementState.Paused:
                    SystemControls.PlaybackStatus = MediaPlaybackStatus.Paused;
                    break;
                case Windows.UI.Xaml.Media.MediaElementState.Playing:
                    SystemControls.PlaybackStatus = MediaPlaybackStatus.Playing;
                    break;
                case Windows.UI.Xaml.Media.MediaElementState.Closed:
                    SystemControls.PlaybackStatus = MediaPlaybackStatus.Closed;
                    break;
                case Windows.UI.Xaml.Media.MediaElementState.Stopped:
                    SystemControls.PlaybackStatus = MediaPlaybackStatus.Stopped;
                    break;
                case Windows.UI.Xaml.Media.MediaElementState.Opening:
                    SystemControls.PlaybackStatus = MediaPlaybackStatus.Changing;
                    break;
            }
        }

        void Player_IsSkipPreviousEnabledChanged(object sender, RoutedEventArgs e)
        {
            RefreshPreviousState();
        }

        private void RefreshPreviousState()
        {
            IsPreviousTrackEnabled = MediaPlayer.InteractiveViewModel != null && MediaPlayer.InteractiveViewModel.IsSkipPreviousEnabled && PreviousTrackExists;
        }

        void Player_IsSkipNextEnabledChanged(object sender, RoutedEventArgs e)
        {
            RefreshNextState();
        }

        private void RefreshNextState()
        {
            IsNextTrackEnabled = MediaPlayer.InteractiveViewModel != null && MediaPlayer.InteractiveViewModel.IsSkipNextEnabled && NextTrackExists;
        }

        void Player_IsPauseEnabledChanged(object sender, RoutedEventArgs e)
        {
            RefreshPauseState();
        }

        private void RefreshPauseState()
        {
            SystemControls.IsPauseEnabled = MediaPlayer.InteractiveViewModel != null;
        }

        void Player_IsPlayResumeEnabledChanged(object sender, RoutedEventArgs e)
        {
            RefreshPlayState();
        }

        private void RefreshPlayState()
        {
            SystemControls.IsPlayEnabled = MediaPlayer.InteractiveViewModel != null;
        }

        void Player_IsFastForwardEnabledChanged(object sender, RoutedEventArgs e)
        {
            RefreshFastForwardState();
        }

        private void RefreshFastForwardState()
        {
            SystemControls.IsFastForwardEnabled = MediaPlayer.InteractiveViewModel != null && MediaPlayer.InteractiveViewModel.IsFastForwardEnabled;
        }

        void Player_IsRewindEnabledChanged(object sender, RoutedEventArgs e)
        {
            RefreshRewindState();
        }

        private void RefreshRewindState()
        {
            SystemControls.IsRewindEnabled = MediaPlayer.InteractiveViewModel != null && MediaPlayer.InteractiveViewModel.IsRewindEnabled;
        }

        void Player_IsStopEnabledChanged(object sender, RoutedEventArgs e)
        {
            RefreshStopState();
        }

        private void RefreshStopState()
        {
            SystemControls.IsStopEnabled = MediaPlayer.InteractiveViewModel != null && MediaPlayer.InteractiveViewModel.IsStopEnabled;
        }

        void MediaPlayer_InteractiveViewModelChanged(object sender, RoutedPropertyChangedEventArgs<IInteractiveViewModel> e)
        {
            if (e.OldValue != null)
            {
                UnwireEvents(e.OldValue);
            }

            RefreshFastForwardState();
            RefreshRewindState();
            RefreshStopState();
            RefreshPlayState();
            RefreshPauseState();
            RefreshNextState();
            RefreshPreviousState();
            RefreshPlaybackStatus();

            if (e.NewValue != null)
            {
                WireEvents(e.NewValue);
            }
        }

        async void SystemControls_ButtonPressed(SystemMediaTransportControls sender, SystemMediaTransportControlsButtonPressedEventArgs args)
        {
            switch (args.Button)
            {
                case SystemMediaTransportControlsButton.Pause:
                    await Dispatcher.RunAsync(Windows.UI.Core.CoreDispatcherPriority.Normal, () =>
                    {
                        if (base.IsActive)
                        {
                            if (MediaPlayer.InteractiveViewModel.IsPauseEnabled)
                            {
                                MediaPlayer.InteractiveViewModel.Pause();
                            }
                        }
                    });
                    break;
                case SystemMediaTransportControlsButton.Play:
                    await Dispatcher.RunAsync(Windows.UI.Core.CoreDispatcherPriority.Normal, () =>
                    {
                        if (base.IsActive)
                        {
                            if (MediaPlayer.InteractiveViewModel.IsPlayResumeEnabled)
                            {
                                MediaPlayer.InteractiveViewModel.PlayResume();
                            }
                        }
                    });
                    break;
                case SystemMediaTransportControlsButton.Stop:
                    await Dispatcher.RunAsync(Windows.UI.Core.CoreDispatcherPriority.Normal, () =>
                    {
                        if (base.IsActive)
                        {
                            if (MediaPlayer.InteractiveViewModel.IsStopEnabled)
                            {
                                MediaPlayer.InteractiveViewModel.Stop();
                            }
                        }
                    });
                    break;
                case SystemMediaTransportControlsButton.Previous:
                    await Dispatcher.RunAsync(Windows.UI.Core.CoreDispatcherPriority.Normal, () =>
                    {
                        if (base.IsActive)
                        {
                            if (isPreviousTrackEnabled && PlaylistPlugin != null)
                            {
                                PlaylistPlugin.GoToPreviousPlaylistItem();
                            }
                        }
                    });
                    break;
                case SystemMediaTransportControlsButton.Next:
                    await Dispatcher.RunAsync(Windows.UI.Core.CoreDispatcherPriority.Normal, () =>
                    {
                        if (base.IsActive)
                        {
                            if (IsNextTrackEnabled && PlaylistPlugin != null)
                            {
                                PlaylistPlugin.GoToNextPlaylistItem();
                            }
                        }
                    });
                    break;
                case SystemMediaTransportControlsButton.Rewind:
                    await Dispatcher.RunAsync(Windows.UI.Core.CoreDispatcherPriority.Normal, () =>
                    {
                        if (base.IsActive)
                        {
                            if (MediaPlayer.InteractiveViewModel.IsRewindEnabled)
                            {
                                MediaPlayer.InteractiveViewModel.DecreasePlaybackRate();
                            }
                        }
                    });
                    break;
                case SystemMediaTransportControlsButton.FastForward:
                    await Dispatcher.RunAsync(Windows.UI.Core.CoreDispatcherPriority.Normal, () =>
                    {
                        if (base.IsActive)
                        {
                            if (MediaPlayer.InteractiveViewModel.IsFastForwardEnabled)
                            {
                                MediaPlayer.InteractiveViewModel.IncreasePlaybackRate();
                            }
                        }
                    });
                    break;
            }
        }
    }
}

File: Universal.Js.Core/js/plugins/SystemTransportControlsPlugin.js
(function (PlayerFramework, undefined) {
    "use strict";

    // SystemTransportControlsPlugin Errors
    var invalidConstruction = "Invalid construction: SystemTransportControlsPlugin constructor must be called using the \"new\" operator.";

    // SystemTransportControlsPlugin Class
    var SystemTransportControlsPlugin = WinJS.Class.derive(PlayerFramework.PluginBase, function (options) {
        if (!(this instanceof PlayerFramework.Plugins.SystemTransportControlsPlugin)) {
            throw invalidConstruction;
        }

        this._isNextTrackEnabled = false;
        this._isPreviousTrackEnabled = false;
        this._observableViewModel = null;
        this._observablePlaylistPlugin = null;
        this.systemControls = null;

        PlayerFramework.PluginBase.call(this, options);
    }, {
        // Private Methods
        _onActivate: function () {
            this.systemControls = Windows.Media.SystemMediaTransportControls.getForCurrentView();
            this._bindEvent("buttonpressed", this.systemControls, this._onButtonPressed);
            this._refreshFastForwardState();
            this._refreshRewindState();
            this._refreshStopState();
            this._refreshPlayState();
            this._refreshPauseState();
            this._refreshNextState();
            this._refreshPreviousState();
            this._refreshPlaybackStatus();

            if (this.mediaPlayer.playlistPlugin) {
                this._observablePlaylistPlugin = WinJS.Binding.as(this.mediaPlayer.playlistPlugin);
                this._bindProperty("currentPlaylistItemIndex", this._observablePlaylistPlugin, this._onCurrentPlaylistItemIndexChanged);
            }

            this._wireEvents();
            this._bindEvent("interactiveviewmodelchange", this.mediaPlayer, this._onMediaPlayerInteractiveViewModelChange);

            return true;
        },

        _onDeactivate: function () {
            if (this._observablePlaylistPlugin) {
                this._unbindProperty("currentPlaylistItemIndex", this._observablePlaylistPlugin, this._onCurrentPlaylistItemIndexChanged);
                this._observablePlaylistPlugin = null;
            }

            this.isNextTrackEnabled = false;
            this.isPreviousTrackEnabled = false;

            this._unbindEvent("interactiveviewmodelchange", this.mediaPlayer, this._onMediaPlayerInteractiveViewModelChange);
            this._unwireEvents();

            this._unbindEvent("buttonpressed", this.systemControls, this._onButtonPressed);
            this.systemControls = null;
        },

        _wireEvents: function () {
            if (this.mediaPlayer.interactiveViewModel) {
                this._observableViewModel = WinJS.Binding.as(this.mediaPlayer.interactiveViewModel);
                this._bindProperty("isPlayResumeDisabled", this._observableViewModel, this._onIsPlayResumeDisabledChanged);
                this._bindProperty("isPauseDisabled", this._observableViewModel, this._onIsPauseDisabledChanged);
                this._bindProperty("isStopDisabled", this._observableViewModel, this._onIsStopDisabledChanged);
                this._bindProperty("isFastForwardDisabled", this._observableViewModel, this._onIsFastForwardDisabledChanged);
                this._bindProperty("isRewindDisabled", this._observableViewModel, this._onIsRewindDisabledChanged);
                this._bindProperty("isSkipPreviousDisabled", this._observableViewModel, this._onIsSkipPreviousDisabledChanged);
                this._bindProperty("isSkipNextDisabled", this._observableViewModel, this._onIsSkipNextDisabledChanged);
                this._bindProperty("state", this._observableViewModel, this._onStateChanged);
            }
        },

        _unwireEvents: function () {
            if (this._observableViewModel) {
                this._unbindProperty("state", this._observableViewModel, this._onStateChanged);
                this._unbindProperty("isPlayResumeDisabled", this._observableViewModel, this._onIsPlayResumeDisabledChanged);
                this._unbindProperty("isPauseDisabled", this._observableViewModel, this._onIsPauseDisabledChanged);
                this._unbindProperty("isStopDisabled", this._observableViewModel, this._onIsStopDisabledChanged);
                this._unbindProperty("isFastForwardDisabled", this._observableViewModel, this._onIsFastForwardDisabledChanged);
                this._unbindProperty("isRewindDisabled", this._observableViewModel, this._onIsRewindDisabledChanged);
                this._unbindProperty("isSkipPreviousDisabled", this._observableViewModel, this._onIsSkipPreviousDisabledChanged);
                this._unbindProperty("isSkipNextDisabled", this._observableViewModel, this._onIsSkipNextDisabledChanged);
                this._observableViewModel = null;
            }
        },

        isPreviousTrackEnabled: {
            get: function () {
                return this._isPreviousTrackEnabled;
            },
            set: function (value) {
                if (this._isPreviousTrackEnabled !== value) {
                    this._isPreviousTrackEnabled = value;
                    this.systemControls.isPreviousEnabled = value;
                }
            }
        },

        isNextTrackEnabled: {
            get: function () {
                return this._isNextTrackEnabled;
            },
            set: function (value) {
                if (this._isNextTrackEnabled !== value) {
                    this._isNextTrackEnabled = value;
                    this.systemControls.isNextEnabled = value;
                }
            }
        },

        nextTrackExists: {
            get: function () {
                return this.mediaPlayer && this.mediaPlayer.playlistPlugin && this.mediaPlayer.playlistPlugin.canGoToNextPlaylistItem();
            }
        },

        previousTrackExists: {
            get: function () {
                return this.mediaPlayer && this.mediaPlayer.playlistPlugin && this.mediaPlayer.playlistPlugin.canGoToPreviousPlaylistItem();
            }
        },

        _onStateChanged: function (e) {
            this._refreshPlaybackStatus();
        },

        _refreshPlaybackStatus: function () {
            switch (this.mediaPlayer.interactiveViewModel.state) {
                case PlayerFramework.ViewModelState.unloaded:
                    this.systemControls.playbackStatus = Windows.Media.MediaPlaybackStatus.closed;
                    break;
                case PlayerFramework.ViewModelState.loading:
                    this.systemControls.playbackStatus = Windows.Media.MediaPlaybackStatus.changing;
                    break;
                case PlayerFramework.ViewModelState.paused:
                    this.systemControls.playbackStatus = Windows.Media.MediaPlaybackStatus.paused;
                    break;
                case PlayerFramework.ViewModelState.playing:
                    this.systemControls.playbackStatus = Windows.Media.MediaPlaybackStatus.playing;
                    break;
            }
        },

        _onMediaPlayerStopped: function (e) {
            this.systemControls.playbackStatus = Windows.Media.MediaPlaybackStatus.stopped;
        },

        _refreshTrackButtonStates: function () {
            this.isNextTrackEnabled = this.nextTrackExists;
            this.isPreviousTrackEnabled = this.previousTrackExists;
        },

        _onCurrentPlaylistItemIndexChanged: function (e) {
            this.isNextTrackEnabled = this.nextTrackExists;
            this.isPreviousTrackEnabled = this.previousTrackExists;
        },

        _onIsSkipPreviousDisabledChanged: function (e) {
            this._refreshPreviousState();
        },

        _refreshPreviousState: function () {
            this.isPreviousTrackEnabled = this.mediaPlayer.interactiveViewModel !== null && !this.mediaPlayer.interactiveViewModel.isSkipPreviousDisabled && this.previousTrackExists;
        },

        _onIsSkipNextDisabledChanged: function (e) {
            this._refreshNextState();
        },

        _refreshNextState: function () {
            this.isNextTrackEnabled = this.mediaPlayer.interactiveViewModel !== null && !this.mediaPlayer.interactiveViewModel.isSkipNextDisabled && this.nextTrackExists;
        },

        _onIsPauseDisabledChanged: function (e) {
            this._refreshPauseState();
        },

        _refreshPauseState: function () {
            this.systemControls.isPauseEnabled = this.mediaPlayer.interactiveViewModel !== null;
        },

        _onIsPlayResumeDisabledChanged: function (e) {
            this._refreshPlayState();
        },

        _refreshPlayState: function () {
            this.systemControls.isPlayEnabled = this.mediaPlayer.interactiveViewModel !== null;
        },

        _onIsFastForwardDisabledChanged: function (e) {
            this._refreshFastForwardState();
        },

        _refreshFastForwardState: function () {
            this.systemControls.isFastForwardEnabled = this.mediaPlayer.interactiveViewModel !== null && !this.mediaPlayer.interactiveViewModel.isFastForwardDisabled;
        },

        _onIsRewindDisabledChanged: function (e) {
            this._refreshRewindState();
        },

        _refreshRewindState: function () {
            this.systemControls.isRewindEnabled = this.mediaPlayer.interactiveViewModel !== null && !this.mediaPlayer.interactiveViewModel.isRewindDisabled;
        },

        _onIsStopDisabledChanged: function (e) {
            this._refreshStopState();
        },

        _refreshStopState: function () {
            this.systemControls.isStopEnabled = this.mediaPlayer.interactiveViewModel !== null && !this.mediaPlayer.interactiveViewModel.isStopDisabled;
        },

        _onMediaPlayerInteractiveViewModelChange: function (e) {
            // rewire to the new VM
            this._unwireEvents();

            this._refreshFastForwardState();
            this._refreshRewindState();
            this._refreshStopState();
            this._refreshPlayState();
            this._refreshPauseState();
            this._refreshNextState();
            this._refreshPreviousState();
            this._refreshPlaybackStatus();

            this._wireEvents();
        },

        _onButtonPressed: function (e) {
            switch (e.button) {
                case Windows.Media.SystemMediaTransportControlsButton.pause:
                    if (!this.mediaPlayer.interactiveViewModel.isPauseDisabled) {
                        this.mediaPlayer.interactiveViewModel.pause();
                    }
                    break;
                case Windows.Media.SystemMediaTransportControlsButton.play:
                    if (!this.mediaPlayer.interactiveViewModel.isPlayResumeDisabled) {
                        this.mediaPlayer.interactiveViewModel.playResume();
                    }
                    break;
                case Windows.Media.SystemMediaTransportControlsButton.stop:
                    if (!this.mediaPlayer.interactiveViewModel.isStopDisabled) {
                        this.mediaPlayer.interactiveViewModel.stop();
                    }
                    break;
                case Windows.Media.SystemMediaTransportControlsButton.previous:
                    this.mediaPlayer.playlistPlugin.goToPreviousPlaylistItem();
                    break;
                case Windows.Media.SystemMediaTransportControlsButton.next:
                    this.mediaPlayer.playlistPlugin.goToNextPlaylistItem();
                    break;
                case Windows.Media.SystemMediaTransportControlsButton.rewind:
                    if (!this.mediaPlayer.interactiveViewModel.isRewindDisabled) {
                        this.mediaPlayer.interactiveViewModel.decreasePlaybackRate();
                    }
                    break;
                case Windows.Media.SystemMediaTransportControlsButton.fastForward:
                    if (!this.mediaPlayer.interactiveViewModel.isFastForwardDisabled) {
                        this.mediaPlayer.interactiveViewModel.increasePlaybackRate();
                    }
                    break;

            }
        }
    });

    // SystemTransportControlsPlugin Mixins
    WinJS.Class.mix(PlayerFramework.MediaPlayer, {
        SystemTransportControlsPlugin: {
            value: null,
            writable: true,
            configurable: true
        }
    });

    // SystemTransportControlsPlugin Exports
    WinJS.Namespace.define("PlayerFramework.Plugins", {
        SystemTransportControlsPlugin: SystemTransportControlsPlugin
    });

})(PlayerFramework);

